File: python/temporal/src/smooth_operator_temporal/temporal.py
"""Temporal-backed durable execution (requires the ``temporalio`` SDK).

An agent turn runs as :class:`AgentTurnWorkflow`, whose side-effects — the model
call and each tool invocation — are Temporal **activities**
(:class:`AgentTurnActivities`). The workflow drives the engine's deterministic
:func:`~smooth_operator_core.drive_turn` orchestration **unchanged** via a
:class:`_WorkflowActivities` adapter that schedules those activities on the
workflow. The in-process path runs the same ``drive_turn`` inline — one loop, two
backends.

This module imports ``temporalio``; importing it therefore requires the optional
``temporal`` extra (``pip install 'smooai-smooth-operator-temporal[temporal]'``),
mirroring how the Rust reference puts all workflow/activity wiring behind an
off-by-default cargo feature. The serde :mod:`.dto` boundary imports no Temporal
SDK and is always available.

Engine handles (the model provider + tool set the activities run against) are held
on the :class:`AgentTurnActivities` **instance** and injected at worker
construction — Python has no free-function-only activity constraint, so this
replaces the Rust reference's process-global ``init_engine`` with plain dependency
injection. The activities reuse the engine's
:class:`~smooth_operator_core.InProcessActivities` as their dispatch surface, so
the durable path's model/tool behavior is identical to the inline loop's.
"""

from __future__ import annotations

import json
import logging
from datetime import timedelta
from typing import Any

from temporalio import activity, workflow

# `drive_turn` is pure/deterministic (no I/O, clock or RNG), so it is safe to run
# as workflow code. Pass its imports (and the engine dispatch surface the
# activities use) through the workflow sandbox unchanged rather than have the
# sandbox re-import the whole engine + `openai`.
with workflow.unsafe.imports_passed_through():
    from smooth_operator_core import (
        AgentOptions,
        InProcessActivities,
        LlmProvider,
        ToolResult,
        TurnPolicy,
        drive_turn,
    )

    from .dto import AgentTurnInput, ModelCallInput, ModelCallOutput, ToolInvokeInput, ToolInvokeOutput

logger = logging.getLogger(__name__)

#: Start-to-close timeout for the model/tool activities, mirroring the Rust
#: reference's 60s default. (Retry-policy tuning lands with the per-step retry work.)
_ACTIVITY_TIMEOUT = timedelta(seconds=60)


class AgentTurnActivities:
    """The side-effecting boundary of a turn, as Temporal activities.

    Holds an :class:`~smooth_operator_core.InProcessActivities` — the engine's own
    dispatch surface (retry/backoff, ``max_tokens`` clamping, tool resolution with
    clearance + hooks) — so the ``model_call`` / ``tool_invoke`` activities behave
    exactly as the inline loop does. Register its bound methods on a
    :class:`temporalio.worker.Worker`.
    """

    def __init__(self, engine: InProcessActivities) -> None:
        self._engine = engine

    @classmethod
    def from_engine(cls, llm: LlmProvider, options: AgentOptions | None = None) -> AgentTurnActivities:
        """Build activities backed by a fresh engine dispatch surface over ``llm``."""
        return cls(InProcessActivities(llm, options or AgentOptions()))

    @activity.defn
    async def health_echo(self, message: str) -> str:
        """Liveness probe backing :class:`HealthWorkflow`."""
        return f"smooth-operator-temporal ok: {message}"

    @activity.defn
    async def model_call(self, input: ModelCallInput) -> ModelCallOutput:
        """The model call — the ``Think`` step — run as a durable activity."""
        assistant = await self._engine.model_call(input.messages, input.tools)
        return ModelCallOutput.from_assistant(assistant)

    @activity.defn
    async def tool_invoke(self, input: ToolInvokeInput) -> ToolInvokeOutput:
        """A single tool invocation — the ``Act`` step — run as a durable activity.

        Tool *business* failures come back on ``is_error`` (not as a raise),
        matching the engine's own dispatch.
        """
        result = await self._engine.tool_invoke(input.call)
        return ToolInvokeOutput(content=result.content, is_error=result.is_error)


def _wait_seconds(call: dict[str, Any]) -> int | None:
    """Extract a non-negative integer ``seconds`` from a wire tool call, or ``None``.

    The wire arguments are a JSON string (the shape the model emits); ``bool`` is
    rejected even though it is an ``int`` subclass in Python.
    """
    raw = call.get("function", {}).get("arguments")
    try:
        args = json.loads(raw) if isinstance(raw, str) else (raw or {})
    except (ValueError, TypeError):
        return None
    seconds = args.get("seconds") if isinstance(args, dict) else None
    if isinstance(seconds, bool) or not isinstance(seconds, int) or seconds < 0:
        return None
    return seconds


class _WorkflowActivities:
    """Makes the engine's ``AgentActivities`` surface schedule Temporal activities.

    ``approval_required`` names tools that block **durably** on an approval signal
    before they run (the HITL unlock — no mid-turn connection state, just a
    signal). ``wait_tool`` (if set) names a built-in **durable wait**: a model call
    to it sleeps the workflow on a Temporal timer, a pause that survives restarts
    and can span days, instead of dispatching an activity.
    """

    def __init__(self, wf: AgentTurnWorkflow, approval_required: list[str], wait_tool: str | None) -> None:
        self._wf = wf
        self._approval_required = set(approval_required)
        self._wait_tool = wait_tool

    async def model_call(self, messages: list[dict[str, Any]], tools: list[dict[str, Any]] | None) -> dict[str, Any]:
        output: ModelCallOutput = await workflow.execute_activity_method(
            AgentTurnActivities.model_call,
            ModelCallInput(messages=messages, tools=tools),
            start_to_close_timeout=_ACTIVITY_TIMEOUT,
        )
        return output.to_assistant()

    async def tool_invoke(self, call: dict[str, Any]) -> ToolResult:
        name = call["function"]["name"]
        call_id = call["id"]

        # Durable wait: sleep on a Temporal timer instead of dispatching an
        # activity. The timer is recorded in workflow history, so the pause
        # survives restarts and can span days.
        if self._wait_tool is not None and name == self._wait_tool:
            seconds = _wait_seconds(call)
            if seconds is None:
                return ToolResult(content=f"wait tool '{name}' requires an integer `seconds` argument", is_error=True)
            await workflow.sleep(timedelta(seconds=seconds))
            return ToolResult(content=f"Waited {seconds}s (durable timer).", is_error=False)

        # Durable human-in-the-loop gate: block until an approve_tool / deny_tool
        # signal names this call id. The wait lives in workflow history, so it
        # survives restarts and can resolve hours later. `drive_turn` is unchanged
        # — the gate lives here.
        if name in self._approval_required:
            await workflow.wait_condition(lambda: call_id in self._wf.approved or call_id in self._wf.denied)
            if call_id in self._wf.denied:
                return ToolResult(content=f"Tool call '{name}' was denied by human approval.", is_error=True)

        output: ToolInvokeOutput = await workflow.execute_activity_method(
            AgentTurnActivities.tool_invoke,
            ToolInvokeInput(call=call),
            start_to_close_timeout=_ACTIVITY_TIMEOUT,
        )
        return ToolResult(content=output.content, is_error=output.is_error)


@workflow.defn
class AgentTurnWorkflow:
    """One agent turn as a Temporal workflow.

    Seeds the conversation, then drives the engine's :func:`~smooth_operator_core.drive_turn`
    unchanged over :class:`_WorkflowActivities`, so the durable path is the same
    loop as in-process. Returns the full message list.

    Workflow state is the human-approval ledger — ``approved`` / ``denied``
    tool-call ids, populated by the ``approve_tool`` / ``deny_tool`` signals and
    observed by the adapter's durable gate.
    """

    def __init__(self) -> None:
        self.approved: set[str] = set()
        self.denied: set[str] = set()

    @workflow.run
    async def run(self, input: AgentTurnInput) -> list[dict[str, Any]]:
        messages: list[dict[str, Any]] = [
            {"role": "system", "content": input.system_prompt},
            {"role": "user", "content": input.user_message},
        ]
        policy = TurnPolicy() if input.max_iterations == 0 else TurnPolicy(max_iterations=input.max_iterations)
        adapter = _WorkflowActivities(self, input.approval_required_tools, input.wait_tool)
        await drive_turn(adapter, messages, input.tools or None, policy)
        return messages

    @workflow.signal
    def approve_tool(self, call_id: str) -> None:
        """A human approves the tool call with this id, unblocking the gate."""
        self.approved.add(call_id)

    @workflow.signal
    def deny_tool(self, call_id: str) -> None:
        """A human denies the tool call with this id; the gate returns a tool-error
        result instead of running it."""
        self.denied.add(call_id)


@workflow.defn
class HealthWorkflow:
    """Scaffold liveness workflow — proves the SDK integrates end to end and backs
    the ephemeral-server integration test."""

    @workflow.run
    async def run(self, message: str) -> str:
        return await workflow.execute_activity_method(
            AgentTurnActivities.health_echo,
            message,
            start_to_close_timeout=timedelta(seconds=10),
        )

File: go/temporal/temporal.go
package temporal

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	core "github.com/SmooAI/smooth-operator-core/go/core"
	"go.temporal.io/sdk/workflow"
)

// Task-queue-independent identifiers used by both the workflow and its consumers.
const (
	// SignalApproveTool names the human-approval signal that unblocks a gated tool
	// call: its payload is the tool-call id to approve.
	SignalApproveTool = "approve_tool"
	// SignalDenyTool names the signal that denies a gated tool call: its payload is
	// the tool-call id to deny (the tool never runs; the model sees an error result).
	SignalDenyTool = "deny_tool"

	// activityModelCall / activityToolInvoke / activityHealthEcho are the registered
	// activity names (defaulting to the AgentTurnActivities method names). The
	// workflow schedules by name so it needs no reference to the activities struct.
	activityModelCall  = "ModelCall"
	activityToolInvoke = "ToolInvoke"
	activityHealthEcho = "HealthEcho"
)

// agentActivityTimeout is the start-to-close timeout for the model/tool activities,
// mirroring the Rust backend's 60s default. (Retry-policy tuning lands with the
// per-step retry work; the SDK default retry policy applies until then.)
const agentActivityTimeout = 60 * time.Second

// AgentTurnActivities is the side-effecting boundary of a durable agent turn: the
// model call and each tool invocation, run as Temporal activities. It delegates to
// the engine's own [core.InProcessActivities] so a durable activity runs the SAME
// side-effect code the in-process path runs inline — the durable and inline paths
// never diverge. Mirrors the Rust `AgentTurnActivities` (which calls
// `engine.llm.chat` / `engine.tools.execute`).
//
// Register it on a worker with `w.RegisterActivity(acts)`; the exported methods are
// registered under their method names (ModelCall / ToolInvoke / HealthEcho).
type AgentTurnActivities struct {
	inproc *core.InProcessActivities
}

// NewAgentTurnActivities builds the activity surface from a model provider, the
// model to call (empty ⇒ the engine default), and the tools available to it.
func NewAgentTurnActivities(llm core.LlmProvider, model string, tools []core.Tool) *AgentTurnActivities {
	return &AgentTurnActivities{inproc: core.NewInProcessActivities(llm, model, tools)}
}

// HealthEcho is a liveness probe backing the scaffold [HealthWorkflow] — it proves
// the SDK integrates end to end without touching the model or tools.
func (*AgentTurnActivities) HealthEcho(_ context.Context, message string) (string, error) {
	return fmt.Sprintf("smooth-operator-temporal ok: %s", message), nil
}

// ModelCall is the "Think" step run as a durable activity. It returns the engine's
// core.ChatResponse directly as the activity output (see dto.go for why no
// projection is needed).
func (a *AgentTurnActivities) ModelCall(ctx context.Context, input ModelCallInput) (core.ChatResponse, error) {
	return a.inproc.ModelCall(ctx, input.Messages, input.Tools)
}

// ToolInvoke is the "Act" step run as a durable activity. Tool BUSINESS failures
// come back on ToolResult.IsError (not as a returned error), matching the engine's
// dispatch — so a business failure is a completed activity whose content goes back
// to the model, while a returned error is retried as a failed activity.
func (a *AgentTurnActivities) ToolInvoke(ctx context.Context, input ToolInvokeInput) (core.ToolResult, error) {
	return a.inproc.ToolInvoke(ctx, input.Call)
}

// AgentTurnInput is the input to [AgentTurnWorkflow]: everything needed to seed the
// conversation and bound the loop. Serializable so it crosses the workflow-start
// boundary. Mirrors the Rust `AgentTurnInput`.
type AgentTurnInput struct {
	// SystemPrompt is the system prompt for the turn (omitted from the seed when empty).
	SystemPrompt string `json:"systemPrompt"`
	// UserMessage is the user message that opens the turn.
	UserMessage string `json:"userMessage"`
	// Tools are the tool specs available to the model.
	Tools []core.ToolSpec `json:"tools,omitempty"`
	// MaxIterations bounds the loop; 0 falls back to the engine's default TurnPolicy.
	MaxIterations int `json:"maxIterations,omitempty"`
	// ApprovalRequiredTools names tools that require human approval before they run.
	// When the model calls one, the workflow blocks DURABLY until an approve_tool /
	// deny_tool signal names that tool call — the block is recorded in workflow
	// history, so it survives worker restarts and can resolve arbitrarily later.
	ApprovalRequiredTools []string `json:"approvalRequiredTools,omitempty"`
	// WaitTool, if set, names a built-in DURABLE WAIT tool. A model call to it with an
	// integer `seconds` argument sleeps the workflow on a Temporal timer (a durable
	// pause that survives restarts and can span days), instead of dispatching a tool
	// activity. Empty disables the wait tool.
	WaitTool string `json:"waitTool,omitempty"`
}

// AgentTurnWorkflow runs one agent turn as a Temporal workflow. It seeds the
// conversation, then drives the engine's [core.DriveTurn] UNCHANGED over a
// workflow-side adapter that schedules the model/tool activities — so the durable
// path is the same loop as in-process. It returns the full conversation (this
// turn's assistant + tool messages appended to the seed).
//
// The human-approval ledger (approved / denied tool-call ids) is workflow state,
// populated by a signal-pump coroutine reading the approve_tool / deny_tool signal
// channels and observed by the adapter's durable gate.
func AgentTurnWorkflow(ctx workflow.Context, input AgentTurnInput) ([]core.ChatMessage, error) {
	adapter := &workflowActivities{
		ctx:              ctx,
		approved:         map[string]bool{},
		denied:           map[string]bool{},
		approvalRequired: toSet(input.ApprovalRequiredTools),
		waitTool:         input.WaitTool,
	}

	// Signal pump: fold approve_tool / deny_tool signals into the ledger as they
	// arrive. Runs on its own workflow coroutine (cooperatively scheduled with the
	// main one — no real concurrency), so map access needs no lock. The coroutine is
	// discarded when the workflow returns.
	workflow.Go(ctx, func(gctx workflow.Context) {
		approveCh := workflow.GetSignalChannel(gctx, SignalApproveTool)
		denyCh := workflow.GetSignalChannel(gctx, SignalDenyTool)
		selector := workflow.NewSelector(gctx)
		selector.AddReceive(approveCh, func(c workflow.ReceiveChannel, _ bool) {
			var id string
			c.Receive(gctx, &id)
			adapter.approved[id] = true
		})
		selector.AddReceive(denyCh, func(c workflow.ReceiveChannel, _ bool) {
			var id string
			c.Receive(gctx, &id)
			adapter.denied[id] = true
		})
		for gctx.Err() == nil {
			selector.Select(gctx)
		}
	})

	messages := make([]core.ChatMessage, 0, 2)
	if input.SystemPrompt != "" {
		messages = append(messages, core.ChatMessage{Role: "system", Content: input.SystemPrompt})
	}
	messages = append(messages, core.ChatMessage{Role: "user", Content: input.UserMessage})

	policy := core.DefaultTurnPolicy()
	if input.MaxIterations > 0 {
		policy.MaxIterations = input.MaxIterations
	}

	// The context.Context handed to DriveTurn is unused by the adapter (which holds
	// the workflow.Context); DriveTurn itself performs no I/O with it.
	return core.DriveTurn(context.Background(), adapter, messages, input.Tools, policy)
}

// HealthWorkflow is a scaffold liveness workflow — it proves the SDK integrates end
// to end and backs the ephemeral-server integration test.
func HealthWorkflow(ctx workflow.Context, message string) (string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{StartToCloseTimeout: 10 * time.Second})
	var echoed string
	if err := workflow.ExecuteActivity(ctx, activityHealthEcho, message).Get(ctx, &echoed); err != nil {
		return "", err
	}
	return echoed, nil
}

// workflowActivities makes the engine's [core.AgentActivities] surface schedule
// Temporal activities on a workflow.Context. It holds the workflow context and the
// approval/wait config + the ledger the signal pump populates.
//
// It intercepts two engine steps before they reach an activity:
//   - the DURABLE WAIT tool → a Temporal timer (workflow.Sleep), recorded in history;
//   - an approval-gated tool → a durable block on the approve/deny ledger.
//
// core.DriveTurn is unchanged; the gates live here.
type workflowActivities struct {
	ctx              workflow.Context
	approved         map[string]bool
	denied           map[string]bool
	approvalRequired map[string]bool
	waitTool         string
}

// ModelCall schedules the model_call activity and returns its response.
func (a *workflowActivities) ModelCall(_ context.Context, messages []core.ChatMessage, tools []core.ToolSpec) (core.ChatResponse, error) {
	var resp core.ChatResponse
	err := workflow.ExecuteActivity(a.activityCtx(), activityModelCall, ModelCallInput{Messages: messages, Tools: tools}).Get(a.ctx, &resp)
	if err != nil {
		return core.ChatResponse{}, fmt.Errorf("model_call activity: %w", err)
	}
	return resp, nil
}

// ToolInvoke handles the durable wait and the human-approval gate, then (for an
// ordinary tool) schedules the tool_invoke activity.
func (a *workflowActivities) ToolInvoke(_ context.Context, call core.ToolCall) (core.ToolResult, error) {
	// Durable wait: sleep on a Temporal timer instead of dispatching an activity.
	if a.waitTool != "" && call.Name == a.waitTool {
		seconds, ok := parseWaitSeconds(call.Arguments)
		if !ok {
			return core.ToolResult{
				ToolCallID: call.ID,
				Content:    fmt.Sprintf("wait tool '%s' requires an integer `seconds` argument", call.Name),
				IsError:    true,
			}, nil
		}
		if err := workflow.Sleep(a.ctx, time.Duration(seconds)*time.Second); err != nil {
			return core.ToolResult{}, fmt.Errorf("durable timer: %w", err)
		}
		return core.ToolResult{
			ToolCallID: call.ID,
			Content:    fmt.Sprintf("Waited %ds (durable timer).", seconds),
			IsError:    false,
		}, nil
	}

	// Durable human-in-the-loop gate: block until an approve_tool / deny_tool signal
	// names this call id. The wait is recorded in workflow history, so it survives
	// restarts and can resolve hours later.
	if a.approvalRequired[call.Name] {
		id := call.ID
		if err := workflow.Await(a.ctx, func() bool { return a.approved[id] || a.denied[id] }); err != nil {
			return core.ToolResult{}, fmt.Errorf("approval wait: %w", err)
		}
		if a.denied[id] {
			return core.ToolResult{
				ToolCallID: call.ID,
				Content:    fmt.Sprintf("Tool call '%s' was denied by human approval.", call.Name),
				IsError:    true,
			}, nil
		}
	}

	var result core.ToolResult
	err := workflow.ExecuteActivity(a.activityCtx(), activityToolInvoke, ToolInvokeInput{Call: call}).Get(a.ctx, &result)
	if err != nil {
		return core.ToolResult{}, fmt.Errorf("tool_invoke activity %q: %w", call.Name, err)
	}
	return result, nil
}

// activityCtx returns the workflow context with the agent activity options applied.
func (a *workflowActivities) activityCtx() workflow.Context {
	return workflow.WithActivityOptions(a.ctx, workflow.ActivityOptions{StartToCloseTimeout: agentActivityTimeout})
}

// parseWaitSeconds reads a non-negative integer `seconds` argument from a tool
// call's raw-JSON arguments. Mirrors the Rust wait tool's `arguments.get("seconds")`.
func parseWaitSeconds(arguments string) (int64, bool) {
	if arguments == "" {
		return 0, false
	}
	var args map[string]any
	if err := json.Unmarshal([]byte(arguments), &args); err != nil {
		return 0, false
	}
	// JSON numbers decode to float64; accept a whole, non-negative value.
	f, ok := args["seconds"].(float64)
	if !ok || f < 0 || f != float64(int64(f)) {
		return 0, false
	}
	return int64(f), true
}

// toSet builds a membership set from a slice (nil-safe).
func toSet(names []string) map[string]bool {
	set := make(map[string]bool, len(names))
	for _, n := range names {
		set[n] = true
	}
	return set
}
